Marlin/src/feature/bltouch.cpp

```diff
@@ -56,41 +46,138 @@ bool BLTouch::triggered() {
+  // Instead of using safe_delay in the following loop, we could use delay
+  // and do the calling of the temperature-manager ourself. See safe_delay() for
+  // details. TODO
+  while (ms > 50) {
+    ms -= 50;
+    safe_delay(50);
+    if (cmd == BLTOUCH_RESET) continue;
+    if (triggered()) return true;
   }
+  safe_delay(ms);
+  return triggered();
+}
 
-  #if ENABLED(BLTOUCH_V3)
-    #if  EITHER(BLTOUCH_FORCE_5V_MODE, ENDSTOPPULLUPS) \
-      || ALL(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN, ENDSTOPPULLUP_ZMIN) \
-      || (USES_Z_MIN_PROBE_ENDSTOP && ENABLED(ENDSTOPPULLUP_ZMIN_PROBE))
-      set_5V_mode();  // Assume 5V DC logic level if endstop pullup resistors are enabled
-    #elif true || ENABLED(BLTOUCH_FORCE_OPEN_DRAIN_MODE)
-      set_OD_mode();
+bool BLTouch::deploy_wrapper() {
+  // Do a DEPLOY
+  #if ENABLED(BLTOUCH_DEBUG_MSGS)
+    SERIAL_ECHOLN("BLTouch DEPLOY requested");
+  #endif
+
+  // Attempt to DEPLOY, wait for DEPLOY_DELAY or ALARM
+  if (_deploy_query_alarm()) {                      
+    // The deploy might have failed or the probe is already triggered (nozzle too low?)
+    #if ENABLED(BLTOUCH_DEBUG_MSGS)
+      SERIAL_ECHOLN("BLTouch ALARM or TRIGGER after DEPLOY, recovering");
     #endif
+
+    clear();                               // Get the probe into start condition
+
+    // Last attempt to DEPLOY      
+    if (_deploy_query_alarm()) {                
+      // The deploy might have failed or the probe is actually triggered (nozzle too low?) again
+      #if ENABLED(BLTOUCH_DEBUG_MSGS)
+        SERIAL_ECHOLN("BLTouch ALARM or TRIGGER after DEPLOY, recovery failed");
+      #endif
+
+      SERIAL_ERROR_MSG(MSG_STOP_BLTOUCH);  // Tell the user something is wrong, needs action
+      stop();                              // but i'ts not too bad, no need to kill, allow restart
+
+      return true;                         // Tell our caller we goofed in case he cares to know
+    }                    
+  }
+
+  // The BLTouch V3 (and V2) and some clones can be used in "SWITCH" mode. When triggered, 
+  // instead of a 10ms pulse, there is an indefinetly long trigger condition, until 
+  // reset by the STOW as soon as this "very very long" pulse is registered. Might help in cases
+  // where the trigger pulse is difficult to register (noise, caps on the signal)
+  // Entering this mode manually also enables M119 display of the probe pin state.
+  #if ENABLED(BLTOUCH_FORCE_SW_MODE) 
+    _set_SW_mode();                        // If explicitely demanded, RESET into SW mode
   #endif
 
-  if (in_deploy) {
-    _deploy();
-    #if ENABLED(BLTOUCH_V3)
-      set_SW_mode();
+  // Now the probe is ready to issue a 10ms pulse (or longer, if in SW mode) when the pin goes up.
+  // The trigger STOW (see motion.cpp for example) will pull up the probes pin as soon as the pulse
+  // is registered.
+
+  if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPAIR("bltouch.deploy_wrapper()");
+
+  return false; // report success to caller
```

Comment:
That would be my preference as well. It would be easy to inverse the logic at all the very few occurences where these wrappers are called. Note that I have seen a few other places where "false" is "success" in Marlin, so I believe that the original coder of bltouch probably thought that he should adhere to it. I just stuck to that logic. Look at the original bltouch.cpp, that's where I took it from.

The callers typically look like this:

if (triggered() & wrapper) -> error exit
... success code...

So it needs a true to error out. 

Saves quite some machine instructions (8-bit code size) was probably the reasoning.

Oherwise it would have been:

if (triggered() & !wrapper) -> error exit
... success code...

One instruction more at least.

Other constructs using if then else are also possible but use even more instructions in otherwise straight line code.

Sooner or later one should identify these places and gradually turn the logic around.

Reply:
> True for command successful, and false for command failed. Thats the typical convention.

In fact, a very common convention is zero for no error, or non-zero for an error code.

Reply:
I didn't write the original, @thinkyhead 
Seen so many places in Marlin where this is "the wrong way around" - correcting them all would give AVR maybe 100 bytes more program code. 32 bit who cares, but AVR?

Reply:
For example, one out of many (`probe.cpp`):
![grafik](https://user-images.githubusercontent.com/5990805/57191358-e0a61f80-6f24-11e9-9cc4-09032b96c98d.png)
It's all over the place. So lecture the original author. But like I said, he liked coding it that way...

Reply:
Here's another one in `motion.cpp`:
![grafik](https://user-images.githubusercontent.com/5990805/57191538-3a0f4e00-6f27-11e9-9bce-950d35d6da59.png)
I just wanted to conform to the prevalent coding style so as not to break anything.